File: client/android/app/src/main/java/org/tinymux/titan/net/TelnetParser.kt
package org.tinymux.titan.net

class TelnetParser(private val sendRaw: (ByteArray) -> Unit) {
    companion object {
        const val IAC: Int = 255; const val DONT: Int = 254; const val DO: Int = 253
        const val WONT: Int = 252; const val WILL: Int = 251; const val SB: Int = 250
        const val GA: Int = 249; const val SE: Int = 240

        const val ECHO: Int = 1; const val SGA: Int = 3; const val TTYPE: Int = 24
        const val NAWS: Int = 31; const val CHARSET: Int = 42
        const val MSSP: Int = 70; const val GMCP: Int = 201
    }

    private enum class State { DATA, IAC, WILL, WONT, DO, DONT, SB, SB_DATA, SB_IAC }

    private var state = State.DATA
    private var sbOption = 0
    private val sbBuf = mutableListOf<Byte>()
    private val lineBuf = StringBuilder()

    var remoteEcho = false; private set
    var nawsAgreed = false; private set
    var nawsWidth = 80; var nawsHeight = 24
    val gmcp = mutableMapOf<String, String>()
    val mssp = mutableMapOf<String, String>()

    var onLine: ((String) -> Unit)? = null
    var onPrompt: ((String) -> Unit)? = null

    fun process(data: ByteArray, offset: Int = 0, length: Int = data.size) {
        for (i in offset until offset + length) {
            val c = data[i].toInt() and 0xFF
            when (state) {
                State.DATA -> if (c == IAC) state = State.IAC else onData(c)
                State.IAC -> when (c) {
                    IAC -> { onData(c); state = State.DATA }
                    WILL -> state = State.WILL; WONT -> state = State.WONT
                    DO -> state = State.DO; DONT -> state = State.DONT
                    SB -> state = State.SB
                    GA -> {
                        if (lineBuf.isNotEmpty()) { onPrompt?.invoke(lineBuf.toString()); lineBuf.clear() }
                        state = State.DATA
                    }
                    else -> state = State.DATA
                }
                State.WILL -> {
                    when (c) {
                        ECHO -> { remoteEcho = true; send(DO, c) }
                        SGA -> send(DO, c)
                        GMCP -> send(DO, c)
                        else -> send(DONT, c)
                    }; state = State.DATA
                }
                State.WONT -> { if (c == ECHO) remoteEcho = false; state = State.DATA }
                State.DO -> {
                    when (c) {
                        NAWS -> { nawsAgreed = true; sendNaws() }
                        TTYPE -> sendTtype()
                        else -> send(WONT, c)
                    }; state = State.DATA
                }
                State.DONT -> { if (c == NAWS) nawsAgreed = false; state = State.DATA }
                State.SB -> { sbOption = c; sbBuf.clear(); state = State.SB_DATA }
                State.SB_DATA -> if (c == IAC) state = State.SB_IAC else sbBuf.add(c.toByte())
                State.SB_IAC -> when (c) {
                    SE -> { handleSubneg(); state = State.DATA }
                    IAC -> { sbBuf.add(0xFF.toByte()); state = State.SB_DATA }
                    else -> state = State.DATA
                }
            }
        }
    }

    private fun onData(c: Int) {
        if (c == 10) {
            val line = lineBuf.toString().trimEnd('\r')
            onLine?.invoke(line)
            lineBuf.clear()
        } else {
            lineBuf.append(c.toChar())
        }
    }

    private fun send(cmd: Int, opt: Int) = sendRaw(byteArrayOf(IAC.toByte(), cmd.toByte(), opt.toByte()))

    fun sendNaws() {
        if (!nawsAgreed) return
        sendRaw(byteArrayOf(
            IAC.toByte(), SB.toByte(), NAWS.toByte(),
            (nawsWidth shr 8).toByte(), (nawsWidth and 0xFF).toByte(),
            (nawsHeight shr 8).toByte(), (nawsHeight and 0xFF).toByte(),
            IAC.toByte(), SE.toByte()
        ))
    }

    private fun sendTtype() {
        val ttype = "Titan-Android"
        val buf = mutableListOf(IAC.toByte(), SB.toByte(), TTYPE.toByte(), 0.toByte())
        ttype.forEach { buf.add(it.code.toByte()) }
        buf.addAll(listOf(IAC.toByte(), SE.toByte()))
        sendRaw(buf.toByteArray())
    }

    fun sendLine(text: String) = sendRaw("$text\r\n".toByteArray(Charsets.UTF_8))

    private fun handleSubneg() {
        val data = sbBuf.toByteArray()
        when (sbOption) {
            TTYPE -> if (data.isNotEmpty() && data[0].toInt() == 1) sendTtype()
            CHARSET -> handleCharset(data)
            GMCP -> {
                val text = String(data, Charsets.UTF_8)
                val sp = text.indexOf(' ')
                if (sp >= 0) gmcp[text.substring(0, sp)] = text.substring(sp + 1)
                else gmcp[text] = ""
            }
            MSSP -> handleMssp(data)
        }
    }

    private fun handleCharset(data: ByteArray) {
        if (data.isEmpty() || data[0].toInt() != 1) return
        val offered = String(data, 1, data.size - 1, Charsets.UTF_8)
        val delim = if (offered.isNotEmpty()) offered[0] else return
        val charsets = offered.substring(1).split(delim)
        val hasUtf8 = charsets.any { it.equals("UTF-8", ignoreCase = true) }
        val resp = mutableListOf(IAC.toByte(), SB.toByte(), CHARSET.toByte())
        if (hasUtf8) { resp.add(2); "UTF-8".forEach { resp.add(it.code.toByte()) } }
        else resp.add(3)
        resp.addAll(listOf(IAC.toByte(), SE.toByte()))
        sendRaw(resp.toByteArray())
    }

    private fun handleMssp(data: ByteArray) {
        var key = ""; var value = ""; var mode = 0
        for (b in data) {
            val v = b.toInt() and 0xFF
            when {
                v == 1 -> { if (mode == 2 && key.isNotEmpty()) mssp[key] = value; key = ""; value = ""; mode = 1 }
                v == 2 -> mode = 2
                mode == 1 -> key += v.toChar()
                mode == 2 -> value += v.toChar()
            }
        }
        if (mode == 2 && key.isNotEmpty()) mssp[key] = value
    }
}

File: client/ios/Titan/Net/TelnetParser.swift
import Foundation

// MARK: - Telnet Constants

enum TelnetCommand: UInt8 {
    case se   = 240  // End of subnegotiation
    case nop  = 241  // No operation
    case dm   = 242  // Data mark
    case brk  = 243  // Break
    case ip   = 244  // Interrupt process
    case ao   = 245  // Abort output
    case ayt  = 246  // Are you there
    case ec   = 247  // Erase character
    case el   = 248  // Erase line
    case ga   = 249  // Go ahead
    case sb   = 250  // Subnegotiation begin
    case will = 251
    case wont = 252
    case `do` = 253
    case dont = 254
    case iac  = 255
}

enum TelnetOption: UInt8 {
    case binary    = 0
    case echo      = 1
    case sga       = 3
    case ttype     = 24
    case naws      = 31
    case charset   = 42
    case compress  = 85   // MCCP v1
    case compress2 = 86   // MCCP v2
    case gmcp      = 201
}

// MARK: - Telnet Parser

struct TelnetParser {
    var onText: ((Data) -> Void)?
    var onLine: ((String) -> Void)?
    var onPrompt: ((String) -> Void)?
    var sendRaw: ((Data) -> Void)?

    var nawsWidth: Int = 80
    var nawsHeight: Int = 24
    var remoteEcho: Bool = false

    private enum State {
        case normal
        case iac
        case will, wont, `do`, dont
        case sb
        case sbData
        case sbIAC
    }

    private var state: State = .normal
    private var textBuffer = Data()
    private var lineBuffer = Data()
    private var subOption: UInt8 = 0
    private var subData = Data()

    // MARK: - Process incoming bytes

    mutating func process(_ data: Data) {
        for byte in data {
            switch state {
            case .normal:
                if byte == TelnetCommand.iac.rawValue {
                    state = .iac
                } else {
                    textBuffer.append(byte)
                    if byte == 0x0A { // newline
                        flushLine()
                    }
                }

            case .iac:
                switch byte {
                case TelnetCommand.iac.rawValue:
                    // Escaped 0xFF — literal byte
                    textBuffer.append(byte)
                    state = .normal
                case TelnetCommand.will.rawValue:
                    state = .will
                case TelnetCommand.wont.rawValue:
                    state = .wont
                case TelnetCommand.do.rawValue:
                    state = .do
                case TelnetCommand.dont.rawValue:
                    state = .dont
                case TelnetCommand.sb.rawValue:
                    state = .sb
                case TelnetCommand.ga.rawValue, 239: // GA or EOR
                    flushPrompt()
                    state = .normal
                case TelnetCommand.nop.rawValue:
                    state = .normal
                default:
                    state = .normal
                }

            case .will:
                handleWill(option: byte)
                state = .normal

            case .wont:
                handleWont(option: byte)
                state = .normal

            case .do:
                handleDo(option: byte)
                state = .normal

            case .dont:
                handleDont(option: byte)
                state = .normal

            case .sb:
                subOption = byte
                subData = Data()
                state = .sbData

            case .sbData:
                if byte == TelnetCommand.iac.rawValue {
                    state = .sbIAC
                } else {
                    subData.append(byte)
                }

            case .sbIAC:
                if byte == TelnetCommand.se.rawValue {
                    handleSubnegotiation(option: subOption, data: subData)
                    state = .normal
                } else if byte == TelnetCommand.iac.rawValue {
                    // Escaped 0xFF within subnegotiation
                    subData.append(byte)
                    state = .sbData
                } else {
                    state = .normal
                }
            }
        }

        // Flush any remaining text
        if !textBuffer.isEmpty {
            onText?(textBuffer)
            textBuffer.removeAll()
        }
    }

    // MARK: - Line handling

    private mutating func flushLine() {
        var line = textBuffer
        textBuffer.removeAll()

        // Strip trailing CR/LF
        while let last = line.last, last == 0x0A || last == 0x0D {
            line.removeLast()
        }

        if let str = String(data: line, encoding: .utf8)
            ?? String(data: line, encoding: .isoLatin1) {
            onLine?(str)
        }
    }

    private mutating func flushPrompt() {
        if textBuffer.isEmpty { return }
        let prompt = textBuffer
        textBuffer.removeAll()

        if let str = String(data: prompt, encoding: .utf8)
            ?? String(data: prompt, encoding: .isoLatin1) {
            onPrompt?(str)
        }
    }

    // MARK: - Option negotiation

    private mutating func handleWill(option: UInt8) {
        switch option {
        case TelnetOption.echo.rawValue:
            remoteEcho = true
            send(command: .do, option: option)
        case TelnetOption.sga.rawValue:
            send(command: .do, option: option)
        case TelnetOption.gmcp.rawValue:
            send(command: .do, option: option)
        default:
            send(command: .dont, option: option)
        }
    }

    private mutating func handleWont(option: UInt8) {
        if option == TelnetOption.echo.rawValue {
            remoteEcho = false
        }
        send(command: .dont, option: option)
    }

    private func handleDo(option: UInt8) {
        switch option {
        case TelnetOption.naws.rawValue:
            send(command: .will, option: option)
            sendNaws()
        case TelnetOption.ttype.rawValue:
            send(command: .will, option: option)
        case TelnetOption.charset.rawValue:
            send(command: .will, option: option)
        default:
            send(command: .wont, option: option)
        }
    }

    private func handleDont(option: UInt8) {
        send(command: .wont, option: option)
    }

    // MARK: - Subnegotiation

    private func handleSubnegotiation(option: UInt8, data: Data) {
        switch option {
        case TelnetOption.ttype.rawValue:
            if data.first == 1 { // SEND
                let name = "Titan".data(using: .ascii)!
                var response = Data([TelnetCommand.iac.rawValue, TelnetCommand.sb.rawValue,
                                     TelnetOption.ttype.rawValue, 0]) // IS
                response.append(name)
                response.append(contentsOf: [TelnetCommand.iac.rawValue, TelnetCommand.se.rawValue])
                sendRaw?(response)
            }

        case TelnetOption.charset.rawValue:
            if data.first == 1 { // REQUEST
                // Accept UTF-8 if offered
                let offered = String(data: data.dropFirst(), encoding: .ascii) ?? ""
                if offered.uppercased().contains("UTF-8") {
                    var response = Data([TelnetCommand.iac.rawValue, TelnetCommand.sb.rawValue,
                                         TelnetOption.charset.rawValue, 2]) // ACCEPTED
                    response.append("UTF-8".data(using: .ascii)!)
                    response.append(contentsOf: [TelnetCommand.iac.rawValue, TelnetCommand.se.rawValue])
                    sendRaw?(response)
                }
            }

        default:
            break
        }
    }

    // MARK: - Send helpers

    private func send(command: TelnetCommand, option: UInt8) {
        sendRaw?(Data([TelnetCommand.iac.rawValue, command.rawValue, option]))
    }

    func sendNaws() {
        var data = Data([TelnetCommand.iac.rawValue, TelnetCommand.sb.rawValue,
                         TelnetOption.naws.rawValue])
        // Width (big-endian, escape 0xFF)
        let wh = UInt8(nawsWidth >> 8)
        let wl = UInt8(nawsWidth & 0xFF)
        let hh = UInt8(nawsHeight >> 8)
        let hl = UInt8(nawsHeight & 0xFF)
        for byte in [wh, wl, hh, hl] {
            data.append(byte)
            if byte == 0xFF { data.append(0xFF) }
        }
        data.append(contentsOf: [TelnetCommand.iac.rawValue, TelnetCommand.se.rawValue])
        sendRaw?(data)
    }

    func sendLine(_ text: String) {
        guard var data = text.data(using: .utf8) else { return }
        data.append(contentsOf: [0x0D, 0x0A])
        sendRaw?(data)
    }
}
